path_selection/path_selection_test.go

```diff
@@ -0,0 +1,162 @@
+	id := [constants.NodeIDLength]byte{}
+	_, err := rand.Reader.Read(id[:])
+	if err != nil {
+		panic(err)
+	}
+	d := pki.MixDescriptor{
+		Name:            name,
+		ID:              id,
+		IsProvider:      isProvider,
+		LoadWeight:      3,
+		TopologyLayer:   uint8(layer),
+		EpochAPublicKey: publicKey,
+		Ipv4Address:     ip,
+		TcpPort:         port,
+	}
+	return &d
+}
+
+func newMixPKI(require *require.Assertions) (pki.Client, map[[constants.NodeIDLength]byte]*ecdh.PrivateKey) {
+	type testDesc struct {
+		Name  string
+		Layer int
+		IP    string
+		Port  int
+	}
+
+	test_providers := []testDesc{
+		{
+			Name:  "acme.com",
+			Layer: 0,
+			IP:    "127.0.0.1",
+			Port:  11240,
+		},
+		{
+			Name:  "nsa.gov",
+			Layer: 0,
+			IP:    "127.0.0.1",
+			Port:  11241,
+		},
+		{
+			Name:  "gchq.uk",
+			Layer: 0,
+			IP:    "127.0.0.1",
+			Port:  11242,
+		},
+		{
+			Name:  "fsb.ru",
+			Layer: 0,
+			IP:    "127.0.0.1",
+			Port:  11243,
+		},
+	}
+
+	test_mixes := []testDesc{
+		{
+			Name:  "nsamix101",
+			Layer: 1,
+			IP:    "127.0.0.1",
+			Port:  11234,
+		},
+		{
+			Name:  "nsamix102",
+			Layer: 2,
+			IP:    "127.0.0.1",
+			Port:  112345,
+		},
+		{
+			Name:  "five_eyes",
+			Layer: 3,
+			IP:    "127.0.0.1",
+			Port:  11236,
+		},
+		{
+			Name:  "gchq123",
+			Layer: 1,
+			IP:    "127.0.0.1",
+			Port:  11237,
+		},
+		{
+			Name:  "fsbspy1",
+			Layer: 2,
+			IP:    "127.0.0.1",
+			Port:  11238,
+		},
+		{
+			Name:  "foxtrot2",
+			Layer: 3,
+			IP:    "127.0.0.1",
+			Port:  11239,
+		},
+	}
+
+	keysMap := make(map[[constants.NodeIDLength]byte]*ecdh.PrivateKey)
+	staticPKI := mix_pki.NewStaticPKI()
+	for _, provider := range test_providers {
+		privKey, err := ecdh.NewKeypair(rand.Reader)
+		require.NoError(err, "ecdh NewKeypair error")
+		descriptor := newMixDescriptor(true, provider.Name, provider.Layer, privKey.PublicKey(), provider.IP, provider.Port)
+		keysMap[descriptor.ID] = privKey
+		staticPKI.ProviderMap[descriptor.Name] = descriptor
+		staticPKI.MixMap[descriptor.ID] = descriptor
+	}
+	for _, mix := range test_mixes {
+		privKey, err := ecdh.NewKeypair(rand.Reader)
+		require.NoError(err, "ecdh NewKeypair error")
+		descriptor := newMixDescriptor(true, mix.Name, mix.Layer, privKey.PublicKey(), mix.IP, mix.Port)
+		keysMap[descriptor.ID] = privKey
+		staticPKI.MixMap[descriptor.ID] = descriptor
+	}
+	return staticPKI, keysMap
+}
+
+func TestPathSelection(t *testing.T) {
+	require := require.New(t)
+	mixPKI, _ := newMixPKI(require)
+	nrHops := 4
+	lambda := float64(.123)
+	factory := New(mixPKI, nrHops, lambda)
+
+	senderProvider := "acme.com"
+	recipientProvider := "nsa.gov"
+	recipientName := "alice"
+	recipientID := [constants.RecipientIDLength]byte{}
+	copy(recipientID[:], []byte(recipientName))
+	forwardRoute, replyRoute, surbID, rtt, err := factory.Build(senderProvider, recipientProvider, recipientID)
+	require.NoError(err, "build route error")
+	require.NotNil(surbID, "surbID should NOT be nil")
```

Comment:
would be great to also have tests for expected error cases - descriptors not being available, invalid paths, etc.

Reply:
indeed! good point.